feat!: collect email addresses of discord users

dckc opened the pull request:
BREAKING CHANGE: disables submission of slogfiles by testnet participants

 - [x] feedback on POST
 - [ ] other fields?

cc @dtribble 

tab dump, for my reference:
 - [Array.prototype.findIndex() - JavaScript | MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex)
 - [google-spreadsheet - npm](https://www.npmjs.com/package/google-spreadsheet)
   - [GoogleSpreadsheetWorksheet getrows](https://theoephraim.github.io/node-google-spreadsheet/#/classes/google-spreadsheet-worksheet?id=fn-getrows)
 - [GridRange  |  Sheets API  |  Google Developers](https://developers.google.com/sheets/api/reference/rest/v4/spreadsheets/other#GridRange)
 - [Discord Developer Portal — Documentation — OAuth2](https://discord.com/developers/docs/topics/oauth2#bots)
 - road not taken:
   - [Inserting, updating, and deleting data using Data Manipulation Language  |  Cloud Spanner  |  Google Cloud](https://cloud.google.com/spanner/docs/dml-tasks#node.js)

dckc committed:
feat: get guild member info details

dckc committed:
feat: access to title of google sheet

dckc committed:
feat: upsert in Google sheet

dckc committed:
feat: discord roles, users

dckc committed:
feat: lookup record in Google sheet

dckc committed:
feat: collect email addresses of discord users (WIP)

WIP because it disables testnet participant submission stuff

dckc committed:
build: lock packages with google-spreadsheet

dckc committed:
feat: use service account for sheets access

dckc committed:
feat: setEmail on POST

dckc committed:
fixup: redirect target after contact form POST

dckc committed:
feat: fall back to user.username for missing nick

dckc committed:
feat: acknowledge submission of contact info

mhofman commented on subm/src/discordGuild.js:
```diff
@@ -104,7 +134,70 @@ const avatarBase = 'https://cdn.discordapp.com/avatars';
+ * @returns { TemplateTag }
+ * @typedef { (parts: TemplateStringsArray, ...args: unknown[]) => string } TemplateTag
+ */
+const makeConfig = env => {
+  return ([name], ..._args) => {
+    const value = env[name];
+    if (value === undefined) {
+      throw Error(`${name} not configured`);
+    }
+    return value;
+  };
+};
+
```
Nit: extract in some shared helpers.

dckc committed:
chore: punt on getting email from discord API, for now

mhofman commented on subm/src/sheetAccess.js:
```diff
@@ -0,0 +1,89 @@
+  const destIndex = range(sheet.rowCount).find(rowIndex => {
+    if (rowIndex === 0) return false; // header row
+    const { value } = sheet.getCell(rowIndex, 0);
+    return key === value || value === null;
```
What is the purpose of `value === null` here? I am concerned that this will return an empty row, and that simultaneous `upsert` calls will pick the same empty row, with a last one win insertion.

dckc replied:
when upsert inserts, the row index is not used. it just calls addRow. But still, the `value === null` does seem left over from an earlier design.

dckc replied:
ugh. that was a bit of a pain to work through. had to fall back from `range(n).find(...)` to a `for` loop.

mhofman commented on subm/src/subm.js:
```diff
@@ -104,24 +118,49 @@ const Site = freeze({
 
   badLogin: () => `${AgoricStyle.top}
-  <p><strong>Login refused.</strong> Only Incentivized Testnet participants are allowed.</p>
+  <p><strong>Login refused.</strong> Only ${keys(Site.testnetRoles).join(
```
Nit: the keys of `testnetRoles` are not exactly the most human friendly.

dckc replied:
I'm not inspired to address that.

mhofman commented on subm/src/subm.js:
```diff
@@ -213,6 +252,36 @@ ${Site.fileList(files)}
+      upsert(sheet, user.id, {
+        userID: user.id,
+        joined_at: member.joined_at,
+        nick: member.nick || (member.user || {}).username || '<nick???>',
```
Save `discriminator` as well?

dckc replied:
the `discriminator` and all the other discord stuff gets saved in the `detail` column

mhofman replied:
👍  totally missed that it includes the whole user object!

mhofman commented on subm/src/discordGuild.js:
```diff
@@ -64,6 +70,15 @@ function getContent(host, path, headers, { get }) {
+ *   username: string,
+ *   discriminator: string,
+ *   avatar?: string,
+ *   email?: string, // ... etc.
```
If we end up capturing email, looks like there is a `verified` flag for it.

dckc replied:
I punted on email from discord for now.

dckc committed:
feat: collect full name etc. as well as email

mhofman commented on subm/src/subm.js:
```diff
@@ -427,23 +521,48 @@ async function main(env, { clock, get, express, makeStorage, admin }) {
+    const detail = await discordAPI.users(user.id);
+    let email = detail.email; // ISSUE: why is email missing? what bot permission am I missing?
+    if (!email) {
+      email = await contact.getEmail();
```
We should probably always fetch this anyway, in case they already updated their email, we shouldn't prepopulate with the Discord email.

mhofman commented on subm/src/subm.js:
```diff
@@ -427,23 +521,48 @@ async function main(env, { clock, get, express, makeStorage, admin }) {
+  });
+  app.post(Site.path.contactForm, loginCheck, async (req, res) => {
+    const contact = /** @type { ReturnType<typeof makeContact> } */ (req.user);
+    const { email } = req.body;
```
I assume body parsing is done safely by the framework?

dckc replied:
yes; see `app.use(express.urlencoded({ extended: true }));` above.

mhofman reviewed (changes requested):
Main concern is concurrent `upsert` using the same row.

dckc committed:
fixup: move ack next to submit button

dckc committed:
refactor: factor out makeConfig

 - prune unused takeWhile

dckc committed:
refactor: use URLSearchParams API

Co-authored-by: Mathieu Hofman <86499+mhofman@users.noreply.github.com>

dckc committed:
refactor: lookup() throws on not found

fp approach to null end-of-data wasn't working.
use a for loop

dckc committed:
fixup: agreeToBeContacted was missing in a few places

dckc committed:
fix: distinguish country from full name field

dckc committed:
feat: encourage completing all fields

 - emphasize acks
   ack: colors from https://isabelcastillo.com/error-info-messages-css

mhofman reviewed (approved):
Looks good.

I find it a little unusual to throw a `RangeError` on no result for the lookup, but it works.

dckc commented:
> I find it a little unusual to throw a `RangeError` on no result for the lookup, ...

This is sort of based on python norms around KeyError, IndexError. Is there some other convention for "not found" in js?

mhofman commented:
> Is there some other convention for "not found" in js?

`-1` :)
